### DISCUSSION TITLE
specifying namespace(s) to filter collection

### DISCUSSION THREAD

**Discussion by peterdnight** [prometheus-operator/kube-prometheus#839] [Q&A]
Desired outcome: **only collect/display metrics for specified namespace(s)**.

I have been able to checkout/build/customize 0.7, but when i add replicas exactly as shown in samples on front page (directly under config), it is ignored. I also tried adding namespaces, and while it seems to affect which roles are being generated - grafana still displays all namespaces.  Given configuration below - is it expected that only the kube-system pods be collected/displayed in grafana?

Hints appreciated.
```
 local monitoringConfiguration =

  (import 'kube-prometheus/kube-prometheus.libsonnet') +
  (import 'kube-prometheus/kube-prometheus-node-ports.libsonnet') +
  
  {
    _config+:: {

    prometheus+:: {
      namespaces+: ['kube-system'],
      replicas: 2   // works if moved to prometheus/prometheus/spec 
    },
   ...
```

**Comment by paulfantom** [comment]
Setting `replicas: 2` is the default and won't have any additional effect.

Following code will result in the same manifests as we have in `manifests/` directory, but prometheus CR would have only 1 replica:
```
local kp =
  (import 'kube-prometheus/kube-prometheus.libsonnet') +
  {
    _config+:: {
      namespace: 'monitoring',

      prometheus+: {
        replicas: 1,
      },
    },
  };

{ ['setup/0namespace-' + name]: kp.kubePrometheus[name] for name in std.objectFields(kp.kubePrometheus) } +
{
  ['setup/prometheus-operator-' + name]: kp.prometheusOperator[name]
  for name in std.filter((function(name) name != 'serviceMonitor'), std.objectFields(kp.prometheusOperator))
} +
{ 'prometheus-operator-serviceMonitor': kp.prometheusOperator.serviceMonitor } +
{ ['node-exporter-' + name]: kp.nodeExporter[name] for name in std.objectFields(kp.nodeExporter) } +
{ ['kube-state-metrics-' + name]: kp.kubeStateMetrics[name] for name in std.objectFields(kp.kubeStateMetrics) } +
{ ['alertmanager-' + name]: kp.alertmanager[name] for name in std.objectFields(kp.alertmanager) } +
{ ['prometheus-' + name]: kp.prometheus[name] for name in std.objectFields(kp.prometheus) } +
{ ['prometheus-adapter-' + name]: kp.prometheusAdapter[name] for name in std.objectFields(kp.prometheusAdapter) } +
{ ['grafana-' + name]: kp.grafana[name] for name in std.objectFields(kp.grafana) }
```

```bash
$ git diff
...
...
diff --git a/manifests/prometheus-prometheus.yaml b/manifests/prometheus-prometheus.yaml
index 0548a95..166acfc 100644
--- a/manifests/prometheus-prometheus.yaml
+++ b/manifests/prometheus-prometheus.yaml
@@ -28,7 +28,7 @@ spec:
   podMonitorSelector: {}
   probeNamespaceSelector: {}
   probeSelector: {}
-  replicas: 2
+  replicas: 1
   resources:
     requests:
       memory: 400Mi
```

**Comment by paulfantom** [comment]
> grafana still displays all namespaces. Given configuration below - is it expected that only the kube-system pods be collected/displayed in grafana?

In prometheus-operator logs do you see ServiceMonitors from other namespaces being picked up? Are targets from other services registered in prometheus `/targets` page?

**Reply by peterdnight** [reply]
thanks for response. In this case - i am attempting to restrict collection of pod metric resources (versus only service monitoring). From your hint above - and reading the docs - it appears namespaces only restricts monitors? To clarify - both promethesius and grafana display ALL namespace compute (collected from metric server).

Use case: support multiple "application tenants" such that each application uses 1-n namespaces; and each has their own grafana and promethesius instance(s). Perhaps this is out of scope of kube promethesius? Current mechanism is multiple kubernetes clusters (fully segregated - each with their own master/workers), which works, but not so efficient.
